### PR TITLE
infra: add support for netdriver fuzzing with honggfuzz

### PR DESCRIPTION
**Pull request by DavidKorczynski**
Adds an example with the mongoose webserver

### CODE REVIEW

File: infra/base-images/base-runner/run_fuzzer
Comment (jonathanmetzman) [inline infra/base-images/base-runner/run_fuzzer:152]: @robertswiecki 
What's the best way to identify honggfuzz target binaries that aren't using persistent mode (not sure if the term is different in honggfuzz, sorry) ?

**Comment by DavidKorczynski** [comment]
I think this could face some of the issues mentioned in https://github.com/google/oss-fuzz/issues/7347

File: infra/base-images/base-runner/run_fuzzer
Comment (DavidKorczynski) [reply infra/base-images/base-runner/run_fuzzer]: Looking for `HonggfuzzNetDriver_main` in the binary worked in my case, run_fuzzer distinguishes between normal honggfuzz drivers and netdriver drivers.

**Comment by robertswiecki** [comment]
Sorry for late response, I think this can be useful

```C
/* HF NetDriver signature - if found within file, it means it's a NetDriver-based binary */
#define _HF_NETDRIVER_SIG "\x01_LIBHFUZZ_NETDRIVER_BINARY_SIGNATURE_\x02\xFF"
```

https://github.com/google/honggfuzz/blob/623d8bb620c0d8ca8ac67a0dea8ab4cf835ea706/honggfuzz.h#L56

**Comment by DavidKorczynski** [comment]
@jonathanmetzman -- you approved this but I wanted to double check if it's ready before merging it in?

**Review by jonathanmetzman — CHANGES_REQUESTED** [review]
Yeah lets wait until I make some ClusterFuzz-side changes to merge this.

File: projects/mongoose/build.sh
Comment (oliverchang) [inline projects/mongoose/build.sh:24]: Can we provide HFUZZ_CC/CXX, LIB_HF_NETDRIVER env vars from the base builder to be used here instead? 

Also, is it possible at all to avoid `hfuzz-clang` and use clang directly here?

**Review by oliverchang — COMMENTED** [review]
We're kicking off a deployment for the ClusterFuzz instance, so this should hopefully be ready to go there soon.

File: projects/mongoose/build.sh
Comment (oliverchang) [reply projects/mongoose/build.sh:24]: This is just coming from how we can simplify integration into upstream build systems. The less custom stuff we need to pass on top of our existing CC/CXX, CFLAGS etc, the better.

File: projects/mongoose/build.sh
Comment (DavidKorczynski) [inline projects/mongoose/build.sh:25]: Remove these two lines

File: projects/mongoose/build.sh
Comment (DavidKorczynski) [reply projects/mongoose/build.sh]: We can avoid hfuzz-clang but we need some macros in there if we do that. I added the logic in the `compile_honggfuzz` so now one has to use `$LIB_HF_NETDRIVER "$HF_ND_CC" and `$LIB_HF_NETDRIVER "$HF_ND_CXX"`
for C and C++ honggfuzz netdriver targets. 

WDYT of the current state?

File: projects/mongoose/build.sh
Comment (oliverchang) [inline projects/mongoose/build.sh:24]: Can we just include this -I flag as part of `HF_NETDRIVER_CFLAGS` (see above comment)?

File: infra/base-images/base-builder/compile_honggfuzz
Comment (oliverchang) [inline infra/base-images/base-builder/compile_honggfuzz:26]: s/ND/NETDRIVER/ for consistency

s/CXX/CXXFLAGS/
s/CC/CFLAGS/

as well so that we refer to these as e.g. HF_NETDRIVER_CFLAGS

Comment (DavidKorczynski) [reply infra/base-images/base-builder/compile_honggfuzz]: I opted to use `HFND_CFLAGS`, `HFND_CXXFLAGS` and `LIB_HFND` instead of `s/ND/NETDRIVER/` as `s/ND/NETDRIVER/` would also change the macro, and the macro is needed as is to be consistent with `hfuzz-clang` (https://github.com/google/honggfuzz/blob/60f01c26d9200a4a1058c35a8fc22441b63fa88c/hfuzz_cc/hfuzz-cc.c#L364-L373), which is needed if we want the fuzzers to be compatible  with honggfuzz netdriver fuzzers from outside OSS-Fuzz.

**Review by oliverchang — APPROVED** [review]
thanks!

**Review by oliverchang — APPROVED** [review]

File: infra/base-images/base-builder/compile_honggfuzz
Comment (jonathanmetzman) [inline infra/base-images/base-builder/compile_honggfuzz:23]: nit: end with period.

File: infra/base-images/base-runner/run_fuzzer
Comment (jonathanmetzman) [inline infra/base-images/base-runner/run_fuzzer:157]: nit: end with period.

File: projects/mongoose/fuzz_netdriver_http.c
Comment (jonathanmetzman) [inline projects/mongoose/fuzz_netdriver_http.c:26]: This will have some issues when we have multiple fuzzers. Because we run bad_build_check on all of them at once but they can't use the same port.

**Review by jonathanmetzman — APPROVED** [review]
lgtm

File: projects/mongoose/fuzz_netdriver_http.c
Comment (DavidKorczynski) [reply projects/mongoose/fuzz_netdriver_http.c:26]: True, would you prefer to address this now? One solution is to scan the fuzzers and check if there is more than one fuzzer with netdriver settings, and, if so, avoid running the fuzzers in parallel.

Comment (robertswiecki) [reply projects/mongoose/fuzz_netdriver_http.c:26]: I'm not sure if that's the case here, but hf netdriver uses net namespaces just for that. It should work with a single port per fuzzing thread.

Comment (jonathanmetzman) [reply projects/mongoose/fuzz_netdriver_http.c:26]: Awesome!

**Comment by jonathanmetzman** [comment]
Trial builds don't work here because they don't work on forked repos :-(

**Comment by DavidKorczynski** [comment]
> Trial builds don't work here because they don't work on forked repos :-(

Do I need to do something to make this work, or how do you prefer to move forward?

**Comment by oliverchang** [comment]
> > Trial builds don't work here because they don't work on forked repos :-(
> 
> Do I need to do something to make this work, or how do you prefer to move forward?

The changes look relatively self contained, so let's merge.

Going forward, for infra-related PRs like these, can you create them off a branch on the main repo instead?

**Comment by jonathanmetzman** [comment]
> > > Trial builds don't work here because they don't work on forked repos :-(
> > 
> > 
> > Do I need to do something to make this work, or how do you prefer to move forward?
> 
> The changes look relatively self contained, so let's merge.
> 
> Going forward, for infra-related PRs like these, can you create them off a branch on the main repo instead?

It shouldn't be too hard to support forks, I'm just avoiding implementing this to reduce bus factor: https://github.com/google/oss-fuzz/issues/8273

**Pull request merged**
